### PR TITLE
Centralize policy calculation in the PolicyRepository

### PR DESCRIPTION
**Pull request by squeed**
Please review commit-by-commit.

This PR attempts to clean up the policy generation logic behind the PolicyRepository interface, reducing the coupling between the endpoint and policy packages. There are no functional changes, just a lot of small cleanups.

The basic outline:

1. remove unused `selectorPolicy` field from the endpoint
2. make the policyCache a simple read-through cache
3. move SelectorPolicy interface to the underlying type, off of the cache placeholder
4. move policy calculation to the PolicyRepository itself
5. unexport the now-internal PolicyCache

### CODE REVIEW

File: pkg/policy/repository.go [hunk 1]
```diff
@@ -124,7 +125,15 @@ type PolicyRepository interface {
 	DeleteByResourceLocked(rid ipcachetypes.ResourceID) (ruleSlice, uint64)
 	GetAuthTypes(localID identity.NumericIdentity, remoteID identity.NumericIdentity) AuthTypes
 	GetEnvoyHTTPRules(l7Rules *api.L7Rules, ns string) (*cilium.HttpNetworkPolicyRules, bool)
-	GetPolicyCache() *PolicyCache
+
+	// GetSelectorPolicy computes the SelectorPolicy for a given identity.
+	//
+	// It returns nil if haveRevision is >= than the already calculated version.
+	// This is used to skip policy calculation when a certain revision delta is
+	// known to not affect the given identity. Pass a haveRevision of 0 to force
```
Comment (tklauser) [inline pkg/policy/repository.go]: The godoc comment mentions `haveRevision` while the parameter is named `skipRevision`. One or the other should be changed to match.

File: pkg/policy/repository.go [hunk 2]
```diff
@@ -930,3 +939,37 @@ func wildcardRule(lbls labels.LabelArray, ingress bool) *rule {
 
 	return r
 }
+
+// GetSelectorPolicy computes the SelectorPolicy for a given identity.
+//
+// It returns nil if skipRevision is >= than the already calculated version.
+// This is used to skip policy calculation when a certain revision delta is
+// known to not affect the given identity. Pass a haveRevision of 0 to force
+// calculation.
```
Comment (tklauser) [inline pkg/policy/repository.go]: This also mentions `haveRevision` while the parameter (and the sentence before) use `skipRevision`.

**Review by tklauser — CHANGES_REQUESTED** [review]
Nice cleanup, some godoc nits inline. Also, the 1st and 4th commit are missing `Signed-off-by` lines.

File: pkg/policy/distillery.go [hunk 3]
```diff
@@ -115,25 +100,24 @@ func (cache *PolicyCache) updateSelectorPolicy(identity *identityPkg.Identity) (
...
+	return cip, true, nil
 }
 
-// LocalEndpointIdentityAdded creates a SelectorPolicy cache entry for the
-// specified Identity, without calculating any policy for it.
+// LocalEndpointIdentityAdded is not needed; we only care about local endpoint
+// deletion
 func (cache *PolicyCache) LocalEndpointIdentityAdded(identity *identityPkg.Identity) {
-	cache.insert(identity)
 }
```
Comment (jrajahalme) [inline pkg/policy/distillery.go:107]: Maybe delete the function then? Looks like this is the only implementation in production code, so maybe this function can be removed from `identitymanager.Observer` interface?

**Review by jrajahalme — APPROVED** [review]
Looks great! Only a nit on what looks like an unnecessary interface function :-)

File: pkg/policy/distillery.go (hunk 3 above)
Comment (squeed) [reply pkg/policy/distillery.go:107]: I actually have a use for this in the very near future, so let's keep it for now.

**Review by tklauser — APPROVED** [review]

**Pull request merged**